pd/src/components/staking.rs

```diff
@@ -29,6 +30,14 @@ pub struct Staking {
     /// persisted at the end of the block for processing at the end of the next
     /// epoch.
     delegation_changes: DelegationChanges,
+    /// New validator definitions accumulated over the course of this block, to be
+    /// persisted at the end of the block for processing at the end of the next
+    /// epoch.
+    new_validator_definitions: ValidatorDefinitions,
+    /// Updates to existing validator definitions accumulated over the course of this block, to be
+    /// persisted at the end of the block for processing at the end of the next
+    /// epoch.
+    updated_validator_definitions: ValidatorDefinitions,
```

Comment:
Do we need to accumulate these over the course of the block, and only add them in at the end?

If we could apply them to the tree state immediately, I think a bunch of our logic would get simpler, because we wouldn't need to, e.g., resolve conflicting validator definitions (we'd just apply the first one, and reject the second one), etc.

There's a comment further down at line 684:
```
        // The new validator definitions can be applied now. We didn't apply them earlier because
        // we don't want rate calculations to apply to newly defined validators.
```
but I'm not sure why we don't want this?

Reply:
Per discussion in discord, let's move this to a future PR.